src/main/scala/com/lambtors/poker_api/module/poker/application/win/GameWinnersFinder.scala

```diff
@@ -0,0 +1,53 @@
+package com.lambtors.poker_api.module.poker.application.win
+
+import scala.concurrent.{ExecutionContext, Future}
+
+import com.lambtors.poker_api.module.poker.domain.{PlayerRepository, PokerGameRepository}
+import com.lambtors.poker_api.module.poker.domain.error.{GameCannotEndWhenRiverIsNotDealt, PokerGameNotFound}
+import com.lambtors.poker_api.module.poker.domain.model._
+
+final class GameWinnersFinder(repository: PokerGameRepository, playerRepository: PlayerRepository)(
+    implicit ec: ExecutionContext) {
+
+  def findWinners(gameId: GameId): Future[List[Player]] = repository.search(gameId).flatMap { gameOpt =>
+    if (gameOpt.isEmpty) {
+      throw PokerGameNotFound(gameId)
+    } else {
+      val game = gameOpt.get
+
+      if (game.tableCards.length < 5) {
+        throw GameCannotEndWhenRiverIsNotDealt(gameId)
+      } else {
+        playerRepository.search(gameId).map { players =>
+          findPlayersWithBestCombination(players, game.tableCards)
+        }
+      }
+    }
+  }
+
+  private def findPlayersWithBestCombination(players: List[Player], tableCards: List[Card]): List[Player] = {
+    val bestCombinations = players.map(findBestCombinationOf(_, tableCards))
+    val highestCardValueOutOfAllCombinations = bestCombinations
+      .map(_.combination match {
+        case HighCard(card) => card.cardValue
+      })
+      .sorted(CardValueOrdering.highestValueToLowest)
+      .head
+    val playerBestCombinationsWithHighestCardValue = bestCombinations.filter(_.combination match {
+      case HighCard(card) => card.cardValue == highestCardValueOutOfAllCombinations
+    })
```

Comment:
same

Reply:
we have to unmask the trait here, so that we can improve this code later

Reply:
we don't know it's a highcard at the trait level, using a match makes it fail (on compile time) when we add an extra win combination, so it's safer